File: spark/spark-streaming/spark-temperature-analysis/src/main/java/org/infinispan/demo/TemperatureAnalysis.java
package org.infinispan.demo;

import java.net.UnknownHostException;
import java.util.Properties;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.storage.StorageLevel;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;
import org.infinispan.client.hotrod.event.ClientEvent;
import org.infinispan.spark.stream.InfinispanJavaDStream;

import scala.Tuple3;

public class TemperatureAnalysis {
    public static final String ISPN_ADDRESS = "127.0.0.1:11222";
    public static final String CACHE_NAME = "default";

    public static void main(String[] args) throws UnknownHostException {
        // Create java spark streaming context
        SparkConf conf = new SparkConf().setAppName("temperature-analysis");
        JavaStreamingContext jssc = new JavaStreamingContext(conf, Durations.seconds(1)); // batch interval of 1 second

        // Create Infinispan stream
        Properties properties = new Properties();
        properties.put("infinispan.client.hotrod.server_list", ISPN_ADDRESS);
        properties.put("infinispan.rdd.cacheName", CACHE_NAME);
        
        JavaDStream<Tuple3<String, Object, ClientEvent.Type>> ispnStream = InfinispanJavaDStream.createInfinispanInputDStream(jssc,
                StorageLevel.MEMORY_ONLY(), properties);

        ispnStream.count().foreach(new Function<JavaRDD<Long>, Void>() {
            public Void call(JavaRDD<Long> rdd) throws Exception {
                System.out.printf("Spark temperature analysis: processed %d data items%n", rdd.first());
                return null;
            }
        });
        
        jssc.start();
        jssc.awaitTermination();
    }
}

File: spark/spark-streaming/spark-temperature-analysis/src/main/scala/org/infinispan/demo/TemperatureAnalysis.scala
package org.infinispan.demo

import java.util.Properties

import org.apache.spark.SparkConf
import org.apache.spark.storage.StorageLevel
import org.apache.spark.streaming.Seconds
import org.apache.spark.streaming.State
import org.apache.spark.streaming.StateSpec
import org.apache.spark.streaming.StreamingContext
import org.apache.spark.streaming.dstream.DStream.toPairDStreamFunctions
import org.infinispan.spark.stream.InfinispanInputDStream

object TemperatureAnalysis {

  def main(args: Array[String]) {

    val sparkConf = new SparkConf().setAppName("TemperatureAnalysis")
    val ssc = new StreamingContext(sparkConf, Seconds(1))
    ssc.checkpoint("/tmp/spark-temperature")

    val config = new Properties
    config.put("infinispan.rdd.cacheName", "default")
    config.put("infinispan.client.hotrod.server_list", "127.0.0.1:11222")
    val ispnStream = new InfinispanInputDStream[String, Double](ssc, StorageLevel.MEMORY_ONLY, config)

    val mapFunc = (key: String, temps: Option[Iterable[Double]], state: State[Map[String, (Double, Long)]]) => {
      val sumMap: Map[String, (Double, Long)] = state.getOption().getOrElse(Map())
      val sums: (Double, Long) = sumMap.getOrElse(key, (0f, 0L))
      val curTemps: Iterable[Double] = temps.getOrElse(List())
      val sumUp = sums._1 + curTemps.sum
      val countUp = sums._2 + curTemps.size
      state.update(sumMap.updated(key, (sumUp, countUp)))
      (key, sumUp/countUp)
    }
    
    val measurementStream = ispnStream.map[Tuple2[String, Double]](rdd => (rdd._1, rdd._2))
    val measurementGrouped = measurementStream.groupByKey()
    val currentAvg = measurementGrouped.mapWithState(StateSpec.function(mapFunc))
    
    currentAvg.foreachRDD(rdd => {
      printf("# items in DStream: %d\n", rdd.count())
      rdd.foreach(item => println("Averages:" + item._1 + " -> " + item._2))
    })
    
    ssc.start()
    ssc.awaitTermination()
  }
}